pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,465 @@
+			return
+		}
+		ciImp, ok = deletedState.Obj.(*mcv1alpha1.ClusterInfoImport)
+		if !ok {
+			klog.ErrorS(nil, "DeletedFinalStateUnknown contains non-ClusterInfoImport object", "object", deletedState.Obj)
+			return
+		}
+	}
+
+	if len(ciImp.Spec.GatewayInfos) == 0 {
+		klog.ErrorS(nil, "Received invalid ClusterInfoImport", "object", obj)
+		return
+	}
+	if net.ParseIP(ciImp.Spec.GatewayInfos[0].GatewayIP) == nil {
+		klog.ErrorS(nil, "Received ClusterInfoImport with invalid Gateway IP", "object", obj)
+		return
+	}
+	c.queue.Add("dummykey")
+}
+
+// Run will create defaultWorkers workers (go routines) which will process
+// the Gateway events from the workqueue.
+func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+	cacheSyncs := []cache.InformerSynced{c.gwListerSynced, c.ciImportListerSynced}
+	klog.InfoS("Starting controller", "controller", controllerName)
+	defer klog.InfoS("Shutting down controller", "controller", controllerName)
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem
+// function in order to read and process a message on the workqueue.
+func (c *MCRouteController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *MCRouteController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if k, ok := obj.(string); !ok {
+		c.queue.Forget(obj)
+		klog.InfoS("Expected string in work queue but got", "object", obj)
+		return true
+	} else if err := c.syncMCFlows(); err == nil {
+		c.queue.Forget(k)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(k)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+	}
+	return true
+}
+
+func (c *MCRouteController) syncMCFlows() error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing flows for Multi-cluster", "time", time.Since(startTime))
+	}()
+	var activeGW *mcv1alpha1.Gateway
+	gws, err := c.gwLister.Gateways(c.namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	activeGW = c.getActiveGateway(gws)
+	if activeGW == nil && c.installedActiveGWName == "" {
+		klog.V(2).InfoS("No active Gateway is found")
+		return nil
+	}
+
+	// Active Gateway Name doesn't change but do a fully flows resync
+	// for any Gateway Spec or ClusterInfoImport changes.
+	klog.InfoS("installed GW", "gateway", c.installedActiveGWName)
+	if activeGW != nil && activeGW.Name == c.installedActiveGWName {
+		if err := c.syncInterClusterFlows(activeGW); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if c.installedActiveGWName != "" {
+		if err := c.cleanupFlows(c.installedActiveGWName); err != nil {
+			return err
+		}
+	}
```

Comment:
Maybe set c.installedActiveGWName to "" here, so even the func returns an error at line 275, we clean up the saved state.